### DISCUSSION TITLE
How to get element frequencies in rules

### DISCUSSION THREAD

**Discussion by ris58h** [antlr/antlr4#4364] [Q&A]
I've managed to list all rules defined in a given grammar but I couldn't find a way to get a frequency of an element inside a rule. I need to know which suffix an element has `?`, `+` or `*`.

```java
public class Generator {
    private String parserGrammarFileName;

    public static void main(String[] args) throws Exception {
        Generator generator = new Generator();
        generator.parserGrammarFileName = "SOME_PATH/JavaScriptParser.g4";
        generator.generate();
    }

    public void generate() throws Exception {
        Tool tool = new Tool();
        Grammar grammar = tool.loadGrammar(parserGrammarFileName);
        if (grammar.getType() == ANTLRParser.LEXER) {
            throw new RuntimeException("Input grammar has LEXER type. Expected type is PARSER or COMBINED.");
        }
        if (grammar.ast.hasErrors) {
            throw new RuntimeException("Grammar errors in " + parserGrammarFileName);
        }
        for (Map.Entry<String, Rule> e : grammar.rules.entrySet()) {
            String ruleName = e.getKey();
            Rule rule = e.getValue();
            Set<String> ruleRefs = getRuleRefs(rule);

            String prefix = rule.isStartRule ? "ROOT " : "";
            System.out.println(prefix + ruleName + ": " + ruleRefs);
        }
    }

    private static Set<String> getRuleRefs(Rule rule) {
        Set<String> refs = new HashSet<>();
        for (int i=1; i<=rule.numberOfAlts; i++) {
            refs.addAll(rule.alt[i].ruleRefs.keySet());
        }
        return refs;
    }
}

```

**Comment by jimidle** [comment]
Perhaps you could explain what you are trying to achieve?

**Reply by ris58h** [reply]
I'm trying to generate classes representing rule nodes. They should have corresponding getters that return nullable, non-nullable or list types.

**Reply by kaby76** [reply]
You need to describe the problem, not show an implementation.

What it sounds like is you want to do is to input an arbitrary grammar and determine whether a parser symbol is nullable (aka "derives the empty string"). I'm not sure what the "list types" of a parser symbol means.

**Reply by ris58h** [reply]
@kaby76 yeah, I had to make it clearer.

For each rule I want to have info about frequencies of it's subrules. I thought that Grammar has such info but it doesn't. I've already achieved this via processing GrammarAST blocks.
